### ISSUE TITLE
Writing few big files to a mounted bucket will eat the same space in temp dir with unclosed file descriptor

### ISSUE THREAD

**Issue by liuk001** [GoogleCloudPlatform/gcsfuse#2644]
**Describe the issue**
We are using gcsfuse version 2.5.1 on Ubuntu 22.04 to stream few big files to a bucket from another application (not local disk).
For few files I mean 5 files for about 500 GiB of data, in different chunks (say 40 + 120 + 130 + 100 + 110).
Files are streamed almost in parallel to the bucket.
We are using a dedicated file system for temp files, and using the option "--temp-dir /cache"
During the process the /cache file system is increasing used space but actually the space is allocated by deleted files,  this is a sample output of "lsof |grep /cache":

gcsfuse   2596 2605 gcsfuse           sysadmin   27u      REG              252,6 50723660288        137 /cache/gcsfuse927132578 (deleted)
gcsfuse   2596 2606 gcsfuse           sysadmin   21u      REG              252,6 52158398976        131 /cache/gcsfuse1015742102 (deleted)
gcsfuse   2596 2606 gcsfuse           sysadmin   22u      REG              252,6 57306098176        132 /cache/gcsfuse1261865104 (deleted)
gcsfuse   2596 2606 gcsfuse           sysadmin   25u      REG              252,6 50938098688        135 /cache/gcsfuse156925053 (deleted)
gcsfuse   2596 2606 gcsfuse           sysadmin   26u      REG              252,6 54890597376        136 /cache/gcsfuse2355972758 (deleted)
gcsfuse   2596 2606 gcsfuse           sysadmin   27u      REG              252,6 50723660288        137 /cache/gcsfuse927132578 (deleted)
gcsfuse   2596 2622 gcsfuse           sysadmin   21u      REG              252,6 52158398976        131 /cache/gcsfuse1015742102 (deleted)
gcsfuse   2596 2622 gcsfuse           sysadmin   22u      REG              252,6 57306098176        132 /cache/gcsfuse1261865104 (deleted)
gcsfuse   2596 2622 gcsfuse           sysadmin   25u      REG              252,6 50938098688        135 /cache/gcsfuse156925053 (deleted)

The /cache file system will fill up to 300-400 GiB.

Thanks in advance.

**System & Version (please complete the following information):**
 - OS: Ubuntu 22.04
 - Platform: GCVE VM
 - Version: gcsfuse version 2.5.1
 
**Steps to reproduce the behavior with following information:**
1. Please share Mount command including all command line or config flags used to mount the bucket.
2. Please make sure you have no other security, monitoring, background processes which can offend the FUSE process running. Possibly reproduce under a fresh/clean installation.
3. Please rerun with --log-severity=TRACE --foreground as additional flags to enable debug logs.
4. Monitor the logs and please capture screenshots or copy the relevant logs to a file (can use --log-format and --log-file as well).
5. Attach the screenshot or the logs file to the bug report here.
6. If you're using gcsfuse with any other library/tool/process please list out the steps you took to reproduce the issue.

Mount command:
gcsfuse --foreground --temp-dir /cache --cache-dir /cache --file-cache-max-size-mb 90000 mybucket $HOME/mybucket
stream to $HOME/mybucket 500GiB in 5 files
look at /cache file system growing to 300-400 GB of occupied space by deleted files
look at deleted files with "lsof |grep /cache"

Looking at file descriptors of the gcsfuse main process:
 ls -la /proc/2596/fd
total 0
dr-x------ 2 sysadmin sysadmin 16 Oct 30 23:31 .
dr-xr-xr-x 9 sysadmin sysadmin  0 Oct 30 23:01 ..
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 0 -> /dev/pts/0
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 1 -> /dev/pts/0
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 12 -> /dev/fuse
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 14 -> 'socket:[41543]'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 2 -> /dev/pts/0
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 21 -> '/cache/gcsfuse1015742102 (deleted)'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 22 -> '/cache/gcsfuse1261865104 (deleted)'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 25 -> '/cache/gcsfuse156925053 (deleted)'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 26 -> '/cache/gcsfuse2355972758 (deleted)'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 27 -> '/cache/gcsfuse927132578 (deleted)'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 3 -> 'socket:[49580]'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 4 -> 'anon_inode:[eventpoll]'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 5 -> 'anon_inode:[eventfd]'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 6 -> 'socket:[56506]'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:31 7 -> 'socket:[57951]'
lrwx------ 1 sysadmin sysadmin 64 Oct 30 23:45 8 -> 'socket:[56507]'

Actually the /cache file system has no files in it, i.e. folder /cache/gcsfuse-file-cache is empty.

**Additional context**
Is it possible that gcsfuse is not properly closing file descriptor of file chunk coming from the input stream?

**SLO:**
We strive to respond to all bug reports within 24 business hours provided the information mentioned above is included.

**Comment by gargnitingoogle** [comment]
@liuk001 apologies for the late response due to the downtime because of holiday season here in India.

TLDR; these files shown as deleted by lsof are anonymous/unlinked files still being used by gcsfuse which will be auto-deleted as soon as their data is uploaded to GCS.

The files showing up as deleted in the temp-dir folder /cache/gcsfuse-file-cache are [anonymous files](https://github.com/jacobsa/fuse/blob/8ccd61173b05c77c16c984a3e951ca402cfa073b/fsutil/fsutil.go#L31), which are unlinked by gcsfuse as soon as they are created, and are auto-deleted as soon as they are done being used. These are temporary files created by gcsfuse for staging the write data for GCS objects, and are persisted until their file handle is closed, or the gcsfuse mount process dies, whichever comes first. How temp-files work (not their implementation) is explained at a high-level in the writes sub-section in the gcsfuse semantics documentation [here](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/docs/semantics.md#readwrites).
. So, these deleted files are actually being used by gcsfuse, but the lsof output is a bit misleading (though technically correct) because gcsfuse has already unlinked them from the file-system for auto-deletion.

> Is it possible that gcsfuse is not properly closing file descriptor of file chunk coming from the input stream?

No I don't think that's the case here. You said that you have 5 files of about 500GiB total size all being written at around the same time, so the temp-dir taking about 300-400 GiB data on disk is expected, because gcsfuse needs to keep the complete data for a file on disk, until it has been completed uploaded to GCS, so the disk usage may even upto complete 500GiB in worst case.

**Comment by gargnitingoogle** [comment]
@liuk001 let me know if the above comment (https://github.com/GoogleCloudPlatform/gcsfuse/issues/2644#issuecomment-2453879224) answers your question.

**Comment by liuk001** [comment]
Hi @gargnitingoogle, and thank you for your feedback.

Yes, your comment answers my question, but unfortunately does not solve my problem :-), since it is not scalable for streaming Terabytes of data directly to the bucket without having a very large local file system for temp files.
I think that the limit is in the FUSE implementation itself.

Actually the streams are coming from the "ovftool" command for exporting virtual machines files from vmware infrastructure, and it can write only to local file system, not directly to S3 natively.

... looking for another tricky way to solve my problem ;-)

Thanks
Luca

**Comment by gargnitingoogle** [comment]
hi @liuk001 . Glad that answered your question. Thanks for explaining your use case in more detail.

> since it is not scalable for streaming Terabytes of data directly to the bucket without having a very large local file system for temp files.

Yes, GCSFuse currently stages every file through a full-size local-file on disk/tmpfs for writing to a GCS object.

However, we have a feature in development already which addresses this problem, where entire file content is not staged locally, but only a window of last N MBs of data is before being uploaded to GCS. Though, I don't know if that'll fully address your problem, or when it'll be released.
cc @vadlakondaswetha.

> ... looking for another tricky way to solve my problem ;-)

I'll ask around in my sister teams to see if there is another client that can help you achieve this in GCS until gcsfuse releases the above feature.

**Comment by gargnitingoogle** [comment]
@liuk001 

As this is a known limitation at this point and it's already being addressed by a feature development which is currently in progress, I am marking this as closed. Please feel free to reopen it if you have any further queries/suggestions.

**Comment by gargnitingoogle** [comment]
@liuk001 I found a potential solution for you.

Streaming uploads: https://cloud.google.com/storage/docs/streaming-uploads#command-line . This allows piping the output of a command to a GCS object. But this needs extra work on your side to create a pipe out of the streamed output.

**Comment by gargnitingoogle** [comment]
As an example of the suggestion in https://github.com/GoogleCloudPlatform/gcsfuse/issues/2644#issuecomment-2459093048, I tested this locally by 
1. creating a fifo file, 
2. writing to it 500 GB data using dd command and then 
3. parallely uploading it to GCS using the streamed output command and it worked for me. 

The following is a log on a machine with 190GB RAM, so it couldn't be storing the whole file in RAM, and I didn't see unusual RAM usage increase either. I checked on disk to see if the dd or gcloud processes were writing the whole content to disk somehow, but they were taking up only about 80 MBs on the disk. So, this pipeline is truly shifting data directly from the producer process (dd) to cloud (gcloud) 

```sh
mkfifo sample_500GB.txt
# in shell 1 (still ongoing)
gcloud storage cp ./sample_500GB.txt gs://$bucket/
Copying file://./sample_500GB.txt to gs://gargnitin-empty-bucket/sample_500GB.txt
:: Completed files 0/1 | 13.0 GB | 15.7 MiB/s
# in shell 2 (still ongoing)
dd if=/dev/urandom of=./sample_500GB.txt bs=1GiB count=500
# parallely in shell 3 (to check file-sizes on disk)
sudo lsof -p 2184224,2184144 | grep sample_500GB.txt
python3 2184144 gargnitin   19r  FIFO  253,1      0t0  83254254 ./sample_500GB.txt
dd      2184224 gargnitin    1w  FIFO  253,1      0t0  83254254 ./sample_500GB.txt
```

**Comment by gargnitingoogle** [comment]
Marking this as closed as @liuk001 has accepted the current status for resolution. Please feel free to reopen in future if needed.
